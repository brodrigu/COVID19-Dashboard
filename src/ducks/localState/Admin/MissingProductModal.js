import { createAction, handleActions } from 'redux-actions';
import { put, takeEvery } from 'redux-saga/effects';
import * as client from '../../../client';
import { addProductMention } from '../../data/content';

// Action creators
export const openModal = createAction('Admin/MissingProductModal/open');
export const closeModal = createAction('Admin/MissingProductModal/close');
export const searchBrand = createAction('Admin/MissingProductModal/searchBrand');
export const searchProduct = createAction('Admin/MissingProductModal/searchProduct');
export const updateMissingProduct = createAction('Admin/MissingProductModal/update');
export const submitMissingProductRequest = createAction('Admin/MissingProductModal/submitMissingProductRequest');
export const submitMissingProductResponse = createAction('Admin/MissingProductModal/submitMissingProductResponse');

const initialState = {
    appearanceSecond: 0,
    brand: {},
    modalIsOpen: false,
    product: {},
    productVariantName: '',
};

// Reducers
export default handleActions({
    [closeModal](state) {
        return { ...state, modalIsOpen: false };
    },
    [openModal](state, { payload }) {
        return {
            ...state,
            appearanceSecond: payload.appearanceSecond,
            modalIsOpen: true,
        };
    },
    [submitMissingProductResponse]() {
        return initialState;
    },
    [updateMissingProduct](state, { payload }) {
        return { ...state, [payload.key]: payload.value };
    },
}, initialState);

export function* workSubmitMissingProductRequest(action) {
    const {
        brand,
        product,
        productVariantName,
        appearanceSecond,
    } = action.payload;
    const data = {
        productVariantName,
    };

    if (brand.id) {
        data.brandId = brand.id;
    } else {
        data.brandName = brand.label.trim();
    }

    if (product.id) {
        data.productId = product.id;
    } else {
        data.productName = product.label.trim();
    }

    const response = yield client.submitMissingProduct(data);
    yield put(submitMissingProductResponse(response));

    yield put(addProductMention({
        appearanceSecond,
        brandName: brand.label,
        missingProductId: response.id,
        productName: product.label,
        productVariantName,
    }));
}

export function* watchSubmitMissingProductRequest() {
    yield takeEvery([submitMissingProductRequest], workSubmitMissingProductRequest);
}
