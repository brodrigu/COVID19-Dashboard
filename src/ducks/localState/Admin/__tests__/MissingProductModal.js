/* eslint-env jest */
import configureStore from 'redux-mock-store';
import { takeEvery } from 'redux-saga/effects';
import * as client from '../../../../client';
// Actions to be tested
import * as MissingProductModalDuck from '../MissingProductModal';

const mockStore = configureStore();
const store = mockStore();


describe('MissingProductModal Actions', () => {
    beforeEach(() => {
        store.clearActions();
    });

    for (const action of [
        'openModal',
        'closeModal',
        'searchBrand',
        'searchProduct',
        'updateMissingProduct',
        'submitMissingProductRequest',
        'submitMissingProductResponse',
    ]) {
        test(`${action} test`, () => {
            store.dispatch(MissingProductModalDuck[action]());
            expect(store.getActions()).toMatchSnapshot();
        });
    }
});

describe('MissingProductModal Reducers', () => {
    beforeEach(() => {
    });

    test('initial state', () => {
        const action = { type: 'dummy_action' };
        expect(MissingProductModalDuck.default(undefined, action)).toMatchSnapshot();
    });

    test('closeModal', () => {
        const action = { type: `${MissingProductModalDuck.closeModal}` };
        expect(MissingProductModalDuck.default({ modalIsOpen: true }, action)).toMatchSnapshot();
    });

    test('openModal', () => {
        const appearanceSecond = 123;
        const action = { payload: appearanceSecond, type: `${MissingProductModalDuck.openModal}` };
        expect(MissingProductModalDuck.default({ modalIsOpen: false }, action)).toMatchSnapshot();
    });

    test('submitMissingProductResponse', () => {
        const action = { type: 'submitMissingProductResponse' };
        expect(MissingProductModalDuck.default(undefined, action)).toMatchSnapshot();
    });

    test('updateMissingProduct', () => {
        const update = { key: 'foo', value: 'bar' };
        const action = { payload: update, type: `${MissingProductModalDuck.updateMissingProduct}` };
        expect(MissingProductModalDuck.default(undefined, action)).toMatchSnapshot();
    });
});


describe('MissingProductModal Sagas', () => {
    beforeEach(() => {
    });

    test('watchSubmitMissingProductRequest', () => {
        const gen = MissingProductModalDuck.watchSubmitMissingProductRequest();
        expect(gen.next().value).toEqual(takeEvery(
            [MissingProductModalDuck.submitMissingProductRequest],
            MissingProductModalDuck.workSubmitMissingProductRequest,
        ));
        expect(gen.next().done).toEqual(true);
    });

    test('workSubmitMissingProductRequest', () => {
        const productWithIds = {
            appearanceSecond: 123,
            brand: { id: 1 },
            product: { id: 2 },
            productVariantName: 'foo',
        };
        const apiResponse = {
            appearanceSecond: productWithIds.appearanceSecond,
            brandName: productWithIds.brand.label,
            missingProductId: 1,
            productName: productWithIds.product.label,
            productVariantName: productWithIds.productVariantName,
        };

        client.submitMissingProduct = jest.fn().mockReturnValue(apiResponse);

        const gen = MissingProductModalDuck.workSubmitMissingProductRequest(
            MissingProductModalDuck.submitMissingProductRequest(productWithIds),
        );
        expect(gen.next().value).toEqual(apiResponse);
        expect(gen.next(apiResponse).value).toMatchSnapshot();
        expect(gen.next().value).toMatchSnapshot();
        expect(gen.next().done).toEqual(true);
    });
});
