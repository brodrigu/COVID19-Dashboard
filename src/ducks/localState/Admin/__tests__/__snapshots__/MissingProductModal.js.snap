// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MissingProductModal Actions closeModal test 1`] = `
Array [
  Object {
    "type": "Admin/MissingProductModal/close",
  },
]
`;

exports[`MissingProductModal Actions openModal test 1`] = `
Array [
  Object {
    "type": "Admin/MissingProductModal/open",
  },
]
`;

exports[`MissingProductModal Actions searchBrand test 1`] = `
Array [
  Object {
    "type": "Admin/MissingProductModal/searchBrand",
  },
]
`;

exports[`MissingProductModal Actions searchProduct test 1`] = `
Array [
  Object {
    "type": "Admin/MissingProductModal/searchProduct",
  },
]
`;

exports[`MissingProductModal Actions submitMissingProductRequest test 1`] = `
Array [
  Object {
    "type": "Admin/MissingProductModal/submitMissingProductRequest",
  },
]
`;

exports[`MissingProductModal Actions submitMissingProductResponse test 1`] = `
Array [
  Object {
    "type": "Admin/MissingProductModal/submitMissingProductResponse",
  },
]
`;

exports[`MissingProductModal Actions updateMissingProduct test 1`] = `
Array [
  Object {
    "type": "Admin/MissingProductModal/update",
  },
]
`;

exports[`MissingProductModal Reducers closeModal 1`] = `
Object {
  "modalIsOpen": false,
}
`;

exports[`MissingProductModal Reducers initial state 1`] = `
Object {
  "appearanceSecond": 0,
  "brand": Object {},
  "modalIsOpen": false,
  "product": Object {},
  "productVariantName": "",
}
`;

exports[`MissingProductModal Reducers openModal 1`] = `
Object {
  "appearanceSecond": undefined,
  "modalIsOpen": true,
}
`;

exports[`MissingProductModal Reducers submitMissingProductResponse 1`] = `
Object {
  "appearanceSecond": 0,
  "brand": Object {},
  "modalIsOpen": false,
  "product": Object {},
  "productVariantName": "",
}
`;

exports[`MissingProductModal Reducers updateMissingProduct 1`] = `
Object {
  "appearanceSecond": 0,
  "brand": Object {},
  "foo": "bar",
  "modalIsOpen": false,
  "product": Object {},
  "productVariantName": "",
}
`;

exports[`MissingProductModal Sagas workSubmitMissingProductRequest 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": Object {
        "appearanceSecond": 123,
        "brandName": undefined,
        "missingProductId": 1,
        "productName": undefined,
        "productVariantName": "foo",
      },
      "type": "Admin/MissingProductModal/submitMissingProductResponse",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`MissingProductModal Sagas workSubmitMissingProductRequest 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": Object {
        "appearanceSecond": 123,
        "brandName": undefined,
        "missingProductId": undefined,
        "productName": undefined,
        "productVariantName": "foo",
      },
      "type": "data/content/addProductMention",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;
